#' Statistical Analysis and Visualization of RSCU Values Across Amino Acids
#'
#' Performs comprehensive statistical analysis and generates publication-quality visualizations
#' of Relative Synonymous Codon Usage (RSCU) values across amino acids. The function conducts:
#' \itemize{
#'   \item Omnibus tests (Kruskal-Wallis, Welch ANOVA, standard ANOVA)
#'   \item Specific tests for 2 codons (Wilcoxon, t-test, Welch's t-test)
#'   \item Post-hoc comparisons (Dunn's test, Pairwise t-tests, Tukey HSD)
#'   \item Multiple testing correction
#' }
#' Results are visualized and saved as PNG files, and a summary table is exported as CSV.
#'
#' @param get_RSCU_out A data frame generated by \code{\link{get_RSCU}} containing RSCU values.
#' @param width Width of the output plots in inches.
#' @param height Height of the output plots in inches.
#' @param res Resolution of the output plots in DPI (dots per inch).
#' @param p.adjust.method Method for p-value adjustment in post-hoc tests. One of
#'   \code{"holm"}, \code{"hochberg"}, \code{"hommel"}, \code{"bonferroni"}, \code{"BH"},
#'   \code{"BY"}, or \code{"fdr"}. Default is \code{"bonferroni"}.
#'
#' @return A data frame with post-hoc statistical results. Side effects:
#'   - PNG files with boxplots, barplots, and scatter plots are saved in the directories \code{boxplots}, \code{barplots}, and \code{scatter_plots}.
#'   - A CSV file \code{Post_hoc_table_aminoacids.csv} is saved in the working directory.
#'
#' @examples
#' withr::with_tempdir({data("prepared_fasta", package = "RSCUcaller")
#' rscu_data <- get_RSCU(prepared_fasta)
#' stat_scat_box(
#'   get_RSCU_out = rscu_data,
#'   width = 6,
#'   height = 6,
#'   res = 300,
#'   p.adjust.method = "bonferroni"
#' )})
#'
#' @seealso
#' \code{\link{get_RSCU}}, \code{\link{boxplot_between_groups}}
#'
#' @importFrom dplyr group_by summarise arrange filter
#' @importFrom magrittr %>%
#' @importFrom rstatix tukey_hsd add_xy_position pairwise_t_test dunn_test levene_test
#' @importFrom ggpubr ggscatter ggboxplot ggbarplot stat_pvalue_manual
#' @importFrom ggplot2 aes geom_jitter theme ggtitle scale_x_discrete element_blank element_text
#' @importFrom grDevices png dev.off
#' @importFrom utils write.csv2
#' @importFrom stats shapiro.test aov oneway.test kruskal.test reformulate t.test wilcox.test
#' @export

stat_scat_box2 <- function(get_RSCU_out, width, height, res, p.adjust.method = "bonferroni") {

  if (base::missing(get_RSCU_out)) {
    stop("The get_RSCU_out predictions are required. Please provide a valid argument.",
         call. = FALSE)
  }

  if (!base::file.exists("scatter_plots") & !base::file.exists("boxplots") & !base::file.exists("barplots")) {
    base::dir.create("scatter_plots")
    base::dir.create("boxplots")
    base::dir.create("barplots")
  }

  if (!(p.adjust.method %in% c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"))) {
    stop('Invalid p.adjust.method  argument. Please choose from "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr" valid options.')
  }

  if (base::file.exists("scatter_plots") & base::file.exists("boxplots") & base::file.exists("barplots")) {

    make_tukey_test <- function(data, variable, grouping_variable) {
      data %>%
        rstatix::tukey_hsd(reformulate(grouping_variable, variable)) %>%
        rstatix::add_xy_position(x = grouping_variable)
    }

    base::message(base::paste0("Calculating statistics and generating plots... "))
    levene_warning_aminoacids <- base::character()
    normality_warning_codons <- base::character()
    aminoacids <- unique(get_RSCU_out$AA) 
    statistical_table <- base::data.frame(row.names = 1, group1 = NA, group2 = NA, p.adj = NA, p.adj.signif = NA, test = NA)

    for (i in 1:base::length(aminoacids)) {
      current_aa <- aminoacids[i]
      table_1 <- get_RSCU_out %>%
        dplyr::filter(AA %in% current_aa) %>%
        dplyr::select(RSCU, codon)

      codons_in_aa <- unique(table_1$codon)

      if (length(codons_in_aa) < 2) {
        message(paste0("Amino acid ", current_aa, " has only one codon. Skipping statistical tests."))
        next
      }

      all_codons_normal <- TRUE
      for (codon_name in codons_in_aa) {
        codon_data <- table_1 %>% dplyr::filter(codon == codon_name)
        if (nrow(codon_data) > 2) { 
          shapiro_test_codon <- try(stats::shapiro.test(codon_data$RSCU), silent = TRUE)
          if (inherits(shapiro_test_codon, "try-error") || shapiro_test_codon$p.value < 0.05) {
            all_codons_normal <- FALSE
            normality_warning_codons <- c(normality_warning_codons, paste0(current_aa, " (", codon_name, ")"))
            break
          }
        } else {
          all_codons_normal <- FALSE
          normality_warning_codons <- c(normality_warning_codons, paste0(current_aa, " (", codon_name, " - not enough data for normality test, assuming non-normal)"))
        }
      }

      test_used <- "Unknown" 
      stat_p <- NA
      post_hoc <- NULL
      post_hoc_x_y <- NULL 

      if (length(codons_in_aa) == 2) {
        if (all_codons_normal) {
          levene_test_result <- rstatix::levene_test(RSCU ~ as.factor(codon), data = table_1)
          levene_p <- levene_test_result$p

          if (!is.na(levene_p) && levene_p >= 0.05) {
            t_test_result <- stats::t.test(RSCU ~ codon, data = table_1, var.equal = TRUE)
            stat_p <- t_test_result$p.value
            test_used <- "Independent_t-test"
            post_hoc_data <- data.frame(
              group1 = codons_in_aa[1],
              group2 = codons_in_aa[2],
              p.adj = stat_p,
              p.adj.signif = ifelse(stat_p < 0.001, "***", ifelse(stat_p < 0.01, "**", ifelse(stat_p < 0.05, "*", "ns"))),
              test = test_used
            )
            post_hoc <- post_hoc_data
            post_hoc_x_y <- rstatix::add_xy_position(post_hoc, x = "group1")
          } else {
            welch_t_test_result <- stats::t.test(RSCU ~ codon, data = table_1, var.equal = FALSE)
            stat_p <- welch_t_test_result$p.value
            test_used <- "Welch_t-test"
            post_hoc_data <- data.frame(
              group1 = codons_in_aa[1],
              group2 = codons_in_aa[2],
              p.adj = stat_p,
              p.adj.signif = ifelse(stat_p < 0.001, "***", ifelse(stat_p < 0.01, "**", ifelse(stat_p < 0.05, "*", "ns"))),
              test = test_used
            )
            post_hoc <- post_hoc_data
            post_hoc_x_y <- rstatix::add_xy_position(post_hoc, x = "group1")
            levene_warning_aminoacids <- c(levene_warning_aminoacids, current_aa)
          }
        } else {
          wilcox_test_result <- stats::wilcox.test(RSCU ~ codon, data = table_1)
          stat_p <- wilcox_test_result$p.value
          test_used <- "Wilcoxon_Rank_Sum"
          post_hoc_data <- data.frame(
            group1 = codons_in_aa[1],
            group2 = codons_in_aa[2],
            p.adj = stat_p,
            p.adj.signif = ifelse(stat_p < 0.001, "***", ifelse(stat_p < 0.01, "**", ifelse(stat_p < 0.05, "*", "ns"))),
            test = test_used
          )
          post_hoc <- post_hoc_data
          post_hoc_x_y <- rstatix::add_xy_position(post_hoc, x = "group1")
        }

      } else {
        if (all_codons_normal) {
          levene_test_result <- rstatix::levene_test(RSCU ~ as.factor(codon), data = table_1)
          levene_p <- levene_test_result$p

          if (!is.na(levene_p) && levene_p >= 0.05) {
            aov_result <- stats::aov(RSCU ~ codon, data = table_1)
            stat_p <- base::summary(aov_result)[[1]]$`Pr(>F)`[1]
            test_used <- "ANOVA"
            post_hoc <- base::as.data.frame(make_tukey_test(data = table_1, variable = "RSCU", grouping_variable = "codon"))
            post_hoc <- dplyr::arrange(.data = post_hoc, p.adj)
            post_hoc$test <- "ANOVA/TukeyHSD"
            post_hoc_x_y <- post_hoc
          } else {
            welch_result <- stats::oneway.test(RSCU ~ codon, data = table_1, var.equal = FALSE)
            stat_p <- welch_result$p.value
            test_used <- "Welch ANOVA"
            post_hoc <- as.data.frame(rstatix::pairwise_t_test(RSCU ~ codon, data = table_1, p.adjust.method = p.adjust.method))
            post_hoc <- dplyr::arrange(.data = post_hoc, p.adj)
            post_hoc_x_y <- rstatix::add_xy_position(post_hoc, x = "codon")
            post_hoc_x_y$test <- "Welch_ANOVA/Pairwise_t_test"
            levene_warning_aminoacids <- c(levene_warning_aminoacids, current_aa)
          }
        } else {
          stat <- stats::kruskal.test(RSCU ~ codon, data = table_1)
          stat_p <- stat$p.value
          post_hoc <- base::as.data.frame(rstatix::dunn_test(RSCU ~ codon, data = table_1, p.adjust.method = p.adjust.method))
          post_hoc <- dplyr::arrange(.data = post_hoc, p.adj)
          post_hoc_x_y <- rstatix::add_xy_position(post_hoc, x = "codon")
          post_hoc_x_y$test <- "Kruskal-Wallis/Dunn"
        }
      }

      if (is.null(post_hoc_x_y)) {
          message(paste0("No post-hoc results to plot for amino acid ", current_aa, ". Skipping plotting."))
          next
      }

      post_hoc_stats <- post_hoc_x_y[, c("group1", "group2", "p.adj", "p.adj.signif", "test")]
      statistical_table <- base::rbind(statistical_table, post_hoc_stats)
      statistical_table <- statistical_table[!base::is.na(statistical_table$group1), ]

      if(length(codons_in_aa) == 2 && !("y.position" %in% names(post_hoc_x_y))) {
        post_hoc_x_y$y.position <- max(table_1$RSCU) * 1.05 
        post_hoc_x_y$groups <- list(codons_in_aa) 
        post_hoc_x_y$xmin <- 1 
        post_hoc_x_y$xmax <- 2 
      } else if (length(codons_in_aa) > 2) {
        post_hoc_x_y <- post_hoc_x_y[, c("group1", "group2", "p.adj", "p.adj.signif", "y.position", "groups", "xmin", "xmax", "test")]
      }


      plot_title <- paste0(current_aa, ", ", test_used, ", p=", round(stat_p, digits = 3))

      cat("The differences between codons for amino acid", current_aa, "had a p-value of:", format(stat_p, digits = 3), "\n")

      png(paste0("scatter_plots/", current_aa, ".png"), width = width, height = height, units = "in", res = res)
      p <- ggpubr::ggscatter(table_1,
                             x = "codon",
                             y = "RSCU",
                             point = "FALSE",
                             palette = c("dodgerblue3", "maroon2", "forestgreen", "darkorange1", "blueviolet", "firebrick2", "darkcyan")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y, label = "p.adj.signif", hide.ns = TRUE) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::geom_jitter(aes(color = codon), width = 0.2, height = 0) +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title))
      base::print(p)
      dev.off()

      png(paste0("boxplots/", current_aa, ".png"), width = width, height = height, units = "in", res = res)
      p <- ggpubr::ggboxplot(table_1,
                             x = "codon",
                             y = "RSCU",
                             fill = "white",
                             color = "codon",
                             palette = c("dodgerblue3", "maroon2", "forestgreen", "darkorange1", "blueviolet", "firebrick2", "darkcyan")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y, label = "p.adj.signif", hide.ns = TRUE) +
        ggplot2::geom_jitter(aes(color = codon), width = 0.2, height = 0) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title))
      base::print(p)
      dev.off()

      table_1_summary <- table_1 %>% dplyr::group_by(codon) %>% dplyr::summarize(`RSCU` = mean(RSCU))
      png(paste0("barplots/", current_aa, ".png"), width = width, height = height, units = "in", res = res)
      p <- ggpubr::ggbarplot(table_1_summary,
                             x = "codon",
                             y = "RSCU",
                             fill = "codon",
                             color = "codon",
                             palette = c("dodgerblue3", "maroon2", "forestgreen", "darkorange1", "blueviolet", "firebrick2", "darkcyan")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y, label = "p.adj.signif", hide.ns = TRUE) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title)) + ggplot2::ylab("mean(RSCU)")
      base::print(p)
      dev.off()
    }

  } else {
    stop("Try to make directory scatter_plots, barplots & boxplots manually and run function again",
         call. = FALSE)
  }
  base::message(base::paste0("Done! Check scatter_plots, barplots & boxplots dirs !!!"))
  write.csv2(statistical_table, "Post_hoc_table_aminoacids.csv", row.names = FALSE)
  return(statistical_table)
}