#' Statistical Analysis and Visualization of RSCU Values Across Amino Acids
#'
#' Performs comprehensive statistical analysis and generates publication-quality visualizations 
#' of Relative Synonymous Codon Usage (RSCU) values across amino acids. The function conducts:
#' \itemize{
#'   \item Omnibus tests (Kruskal-Wallis, Welch ANOVA, standard ANOVA)
#'   \item Post-hoc comparisons (Dunn's test, Pairwise t-tests, Tukey HSD)
#'   \item Multiple testing correction
#' }
#' Results are visualized and saved as PNG files, and a summary table is exported as CSV.
#'
#' @param get_RSCU_out A data frame generated by \code{\link{get_RSCU}} containing RSCU values.
#' @param width Width of the output plots in inches.
#' @param height Height of the output plots in inches.
#' @param res Resolution of the output plots in DPI (dots per inch).
#' @param p.adjust.method Method for p-value adjustment in post-hoc tests. One of
#'   \code{"holm"}, \code{"hochberg"}, \code{"hommel"}, \code{"bonferroni"}, \code{"BH"},
#'   \code{"BY"}, or \code{"fdr"}. Default is \code{"bonferroni"}.
#'
#' @return A data frame with post-hoc statistical results. Side effects: 
#'   - PNG files with boxplots, barplots, and scatter plots are saved in the directories \code{boxplots}, \code{barplots}, and \code{scatter_plots}.
#'   - A CSV file \code{Post_hoc_table_aminoacids.csv} is saved in the working directory.
#'
#' @examples
#' withr::with_tempdir({data("prepared_fasta", package = "RSCUcaller")
#' rscu_data <- get_RSCU(prepared_fasta)
#' stat_scat_box(
#'   get_RSCU_out = rscu_data,
#'   width = 6,
#'   height = 6,
#'   res = 300,
#'   p.adjust.method = "bonferroni"
#' )})
#'
#' @seealso
#' \code{\link{get_RSCU}}, \code{\link{boxplot_between_groups}}
#'
#' @importFrom dplyr group_by summarise arrange
#' @importFrom magrittr %>%
#' @importFrom rstatix tukey_hsd add_xy_position pairwise_t_test dunn_test levene_test
#' @importFrom ggpubr ggscatter ggboxplot ggbarplot stat_pvalue_manual
#' @importFrom ggplot2 aes geom_jitter theme ggtitle scale_x_discrete element_blank element_text
#' @importFrom grDevices png dev.off
#' @importFrom utils write.csv2
#' @importFrom stats shapiro.test aov oneway.test kruskal.test reformulate
#' @export

stat_scat_box <-function(get_RSCU_out,width,height,res,p.adjust.method = "bonferroni") {
  
  if (base::missing(get_RSCU_out)) {
    stop("The get_RSCU_out predictions are required. Please provide a valid argument.",
         call. = FALSE)
  }
  
  if (!base::file.exists("scatter_plots") & !base::file.exists("boxplots") & !base::file.exists("barplots")){
    base::dir.create("scatter_plots")
    base::dir.create("boxplots")
    base::dir.create("barplots")
  }
  
  if (!(p.adjust.method %in% c("holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr"))) {
    stop('Invalid p.adjust.method  argument. Please choose from "holm", "hochberg", "hommel", "bonferroni", "BH", "BY", "fdr" valid options.')
  }
  
  if (base::file.exists("scatter_plots") & base::file.exists("boxplots") & base::file.exists("barplots")) {
    
    make_tukey_test <- function (data,variable,grouping_variable){
      data %>%
        rstatix::tukey_hsd(reformulate(grouping_variable, variable)) %>%
        rstatix::add_xy_position(x = grouping_variable)
    }
    
    base::message(base::paste0("Calculating statistics and generating plots... "))
    levene_warning_codons <- base::character()
    normality_warning_codons <- base::character()
    aminoacids <- c("Ala","Arg","Asn","Asp","Cys","Gln","Glu","Gly","His","Ile","Leu","Lys","Phe","Pro","Ser","Stp","Thr","Tyr","Val")
    statistical_table <- base::data.frame(row.names = 1, group1 = NA, group2 = NA, p.adj=NA, p.adj.signif=NA,test=NA)
    
    for (i in 1:base::length(aminoacids)) {
      table_1 <- get_RSCU_out[get_RSCU_out$AA %in% aminoacids[i],]
      table_1 <- table_1[,c("RSCU", "codon")]
      
      test_used <- "Kruskal-Wallis"
      stat_p <- NA
      post_hoc <- NULL
      
      normality_check_passed <- TRUE
      for (codon_val in unique(table_1$codon)) {
        codon_data <- table_1$RSCU[table_1$codon == codon_val]
        if (length(codon_data) > 3) {
          shapiro_test <- try(stats::shapiro.test(codon_data), silent = TRUE)
          if (inherits(shapiro_test, "try-error") || shapiro_test$p.value < 0.05) {
            normality_check_passed <- FALSE
            break
          }
        } else {
          normality_check_passed <- FALSE
          break
        }
      }
      
      if (normality_check_passed) {
        levene_test <- rstatix::levene_test(RSCU ~ as.factor(codon),
                                            data = table_1)
        levene_p <- levene_test$p
        if (!is.na(levene_p) && levene_p >= 0.05) {
          aov_result <- stats::aov(RSCU ~ codon, data = table_1)
          stat_p <- base::summary(aov_result)[[1]]$`Pr(>F)`[1]
          test_used <- "ANOVA"
          post_hoc <- base::as.data.frame(make_tukey_test(data = table_1,
                                                          variable = "RSCU ", grouping_variable = "codon"))
          post_hoc <- dplyr::arrange(.data = post_hoc,
                                     p.adj)
          post_hoc$test <- "ANOVA/TukeyHSD"
          post_hoc_x_y <- post_hoc
        }
        else {
          welch_result <- stats::oneway.test(RSCU ~
                                               codon, data = table_1, var.equal = FALSE)
          stat_p <- welch_result$p.value
          test_used <- "Welch ANOVA"
          post_hoc <- as.data.frame(rstatix::pairwise_t_test(RSCU ~
                                                               codon, data = table_1, p.adjust.method = p.adjust.method))
          post_hoc <- dplyr::arrange(.data = post_hoc,
                                     p.adj)
          post_hoc_x_y <- rstatix::add_xy_position(post_hoc,
                                                   x = post_hoc[2])
          post_hoc_x_y$test <- "Welch_ANOVA/Pairwise_t_test"
          levene_warning_codons <- c(levene_warning_codons,
                                     aminoacids[i])
        }
      }
      else {
        normality_warning_codons <- c(normality_warning_codons,
                                      aminoacids[i])
      }
      if (!normality_check_passed) {
        stat <- stats::kruskal.test(RSCU ~ codon, data = table_1)
        stat_p <- stat$p.value
        post_hoc <- base::as.data.frame(rstatix::dunn_test(RSCU ~
                                                             codon, data = table_1, p.adjust.method = p.adjust.method))
        post_hoc <- dplyr::arrange(.data = post_hoc,
                                   p.adj)
        post_hoc_x_y <- rstatix::add_xy_position(post_hoc,
                                                 x = post_hoc[2])
        post_hoc_x_y$test <- "Kruskal-Wallis/Dunn"
      }
      post_hoc_stats <- post_hoc_x_y[, c("group1", "group2",
                                         "p.adj", "p.adj.signif", "test")]
      statistical_table <- base::rbind(statistical_table,
                                       post_hoc_stats)
      statistical_table <- statistical_table[!base::is.na(statistical_table$group1),
      ]
      post_hoc_x_y <- post_hoc_x_y[, c("group1", "group2",
                                       "p.adj", "p.adj.signif", "y.position", "groups",
                                       "xmin", "xmax", "test")]
      plot_title <- paste0(aminoacids[i], ", ", test_used,
                           ", p=", format(stat_p, digits = 3))
      
      cat("The differences between codons in the amino acid",aminoacids[i],"had a pvalue of:",stat_p,"\n") # Zmieniono stat$p.value na stat_p
      
      png(paste0("scatter_plots/",aminoacids[i],".png"), width=width, height=height, units = "in", res=res)
      p <- ggpubr::ggscatter(table_1,
                             x="codon",
                             y="RSCU",
                             point = "FALSE",
                             palette = c("dodgerblue3", "maroon2",  "forestgreen", "darkorange1", "blueviolet", "firebrick2")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y, label="p.adj.signif", hide.ns=TRUE) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::geom_jitter(aes(color = codon), width = 0.2, height = 0) +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title))
      base::print(p)
      dev.off()
      png(paste0("boxplots/",aminoacids[i],".png"), width=width, height=height, units = "in", res = res)
      p <- ggpubr::ggboxplot(table_1,
                             x="codon",
                             y="RSCU",
                             fill = "white",
                             color = "codon",
                             palette = c("dodgerblue3", "maroon2",  "forestgreen", "darkorange1", "blueviolet", "firebrick2")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y, label="p.adj.signif", hide.ns=TRUE) +
        ggplot2::geom_jitter(aes(color = codon), width = 0.2, height = 0) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title))
      base::print(p)
      dev.off()
      table_1 <- table_1 %>% dplyr::group_by(codon) %>% dplyr::summarize(`RSCU` = mean(RSCU))
      png(paste0("barplots/", aminoacids[i], ".png"), width = width, height = height, units = "in", res = res)
      p <- ggpubr::ggbarplot(table_1,
                             x = "codon",
                             y = "RSCU",
                             fill = "codon",
                             color = "codon",
                             palette = c("dodgerblue3", "maroon2", "forestgreen", "darkorange1", "blueviolet", "firebrick2")) +
        ggpubr::stat_pvalue_manual(post_hoc_x_y,label = "p.adj.signif", hide.ns = TRUE) +
        ggplot2::theme(legend.position = "none") +
        ggplot2::scale_x_discrete() +
        ggplot2::ggtitle(base::paste0(plot_title)) + ggplot2::ylab("mean(RSCU)")
      base::print(p)
      dev.off()
    }
  }
  else (
    stop("Try to make directory scatter_plots, barplots & boxplots manually and run function again",
         call. = FALSE)
  )
  base::message(base::paste0("Done! Check scatter_plots, barplots & boxplots dirs !!!"))
  write.csv2(statistical_table,"Post_hoc_table_aminoacids.csv",row.names = FALSE)
  return(statistical_table)
}
